File: packages/dartcv/lib/src/video/video.dart
// Copyright (c) 2024, rainyl and all contributors. All rights reserved.
// Use of this source code is governed by a Apache-2.0 license
// that can be found in the LICENSE file.

library cv.video;

import 'dart:ffi' as ffi;

import 'package:ffi/ffi.dart';

import '../core/base.dart';
import '../core/mat.dart';
import '../core/mat_type.dart';
import '../core/point.dart';
import '../core/rect.dart';
import '../core/size.dart';
import '../core/termcriteria.dart';
import '../core/vec.dart';
import '../g/constants.g.dart';
import '../g/video.g.dart' as cvg;
import '../native_lib.dart' show cvideo;

class BackgroundSubtractorMOG2 extends CvStruct<cvg.BackgroundSubtractorMOG2> {
  BackgroundSubtractorMOG2(cvg.BackgroundSubtractorMOG2Ptr ptr, [bool attach = true])
      : super.fromPointer(ptr) {
    if (attach) {
      finalizer.attach(this, ptr.cast(), detach: this);
    }
  }
  factory BackgroundSubtractorMOG2.empty() {
    final p = calloc<cvg.BackgroundSubtractorMOG2>();
    cvRun(() => cvideo.cv_BackgroundSubtractorMOG2_create(p));
    return BackgroundSubtractorMOG2(p);
  }
  factory BackgroundSubtractorMOG2.create({
    int history = 500,
    double varThreshold = 16,
    bool detectShadows = true,
  }) {
    final p = calloc<cvg.BackgroundSubtractorMOG2>();
    cvRun(
      () => cvideo.cv_BackgroundSubtractorMOG2_create_1(
        history,
        varThreshold,
        detectShadows,
        p,
      ),
    );
    return BackgroundSubtractorMOG2(p);
  }

  Mat apply(Mat src) {
    final dst = Mat.empty();
    cvRun(() => cvideo.cv_BackgroundSubtractorMOG2_apply(ref, src.ref, dst.ref, ffi.nullptr));
    return dst;
  }

  @override
  cvg.BackgroundSubtractorMOG2 get ref => ptr.ref;
  static final finalizer =
      OcvFinalizer<cvg.BackgroundSubtractorMOG2Ptr>(cvideo.addresses.cv_BackgroundSubtractorMOG2_close);

  void dispose() {
    finalizer.detach(this);
    cvideo.cv_BackgroundSubtractorMOG2_close(ptr);
  }
}

class BackgroundSubtractorKNN extends CvStruct<cvg.BackgroundSubtractorKNN> {
  BackgroundSubtractorKNN(cvg.BackgroundSubtractorKNNPtr ptr, [bool attach = true]) : super.fromPointer(ptr) {
    if (attach) {
      finalizer.attach(this, ptr.cast(), detach: this);
    }
  }

  static final finalizer =
      OcvFinalizer<cvg.BackgroundSubtractorKNNPtr>(cvideo.addresses.cv_BackgroundSubtractorKNN_close);

  void dispose() {
    finalizer.detach(this);
    cvideo.cv_BackgroundSubtractorKNN_close(ptr);
  }

  factory BackgroundSubtractorKNN.empty() {
    final p = calloc<cvg.BackgroundSubtractorKNN>();
    cvRun(() => cvideo.cv_BackgroundSubtractorKNN_create(p));
    return BackgroundSubtractorKNN(p);
  }
  factory BackgroundSubtractorKNN.create({
    int history = 500,
    double varThreshold = 16,
    bool detectShadows = true,
  }) {
    final p = calloc<cvg.BackgroundSubtractorKNN>();
    cvRun(
      () => cvideo.cv_BackgroundSubtractorKNN_create_1(
        history,
        varThreshold,
        detectShadows,
        p,
      ),
    );
    return BackgroundSubtractorKNN(p);
  }

  Mat apply(Mat src) {
    final dst = Mat.empty();
    cvRun(() => cvideo.cv_BackgroundSubtractorKNN_apply(ref, src.ref, dst.ref, ffi.nullptr));
    return dst;
  }

  @override
  cvg.BackgroundSubtractorKNN get ref => ptr.ref;
}

/// NewBackgroundSubtractorMOG2 returns a new BackgroundSubtractor algorithm
/// of type MOG2. MOG2 is a Gaussian Mixture-based Background/Foreground
/// Segmentation Algorithm.
///
/// For further details, please see:
/// https://docs.opencv.org/master/de/de1/group__video__motion.html#ga2beb2dee7a073809ccec60f145b6b29c
/// https://docs.opencv.org/master/d7/d7b/classcv_1_1BackgroundSubtractorMOG2.html
BackgroundSubtractorMOG2 createBackgroundSubtractorMOG2({
  int history = 500,
  double varThreshold = 16,
  bool detectShadows = true,
}) {
  final p = calloc<cvg.BackgroundSubtractorMOG2>();
  cvRun(
    () => cvideo.cv_BackgroundSubtractorMOG2_create_1(
      history,
      varThreshold,
      detectShadows,
      p,
    ),
  );
  return BackgroundSubtractorMOG2(p);
}

/// Apply computes a foreground mask using the current BackgroundSubtractorMOG2.
///
/// For further details, please see:
/// https://docs.opencv.org/master/d7/df6/classcv_1_1BackgroundSubtractor.html#aa735e76f7069b3fa9c3f32395f9ccd21

/// CalcOpticalFlowFarneback computes a dense optical flow using
/// Gunnar Farneback's algorithm.
///
/// For further details, please see:
/// https://docs.opencv.org/master/dc/d6b/group__video__track.html#ga5d10ebbd59fe09c5f650289ec0ece5af
Mat calcOpticalFlowFarneback(
  InputArray prev,
  InputArray next,
  InputOutputArray flow,
  double pyrScale,
  int levels,
  int winsize,
  int iterations,
  int polyN,
  double polySigma,
  int flags,
) {
  cvRun(
    () => cvideo.cv_calcOpticalFlowFarneback(
      prev.ref,
      next.ref,
      flow.ref,
      pyrScale,
      levels,
      winsize,
      iterations,
      polyN,
      polySigma,
      flags,
      ffi.nullptr,
    ),
  );
  return flow;
}

/// CalcOpticalFlowPyrLK calculates an optical flow for a sparse feature set using
/// the iterative Lucas-Kanade method with pyramids.
///
/// For further details, please see:
/// https://docs.opencv.org/master/dc/d6b/group__video__track.html#ga473e4b886d0bcc6b65831eb88ed93323
(VecPoint2f nextPts, VecUChar? status, VecF32? error) calcOpticalFlowPyrLK(
  InputArray prevImg,
  InputArray nextImg,
  VecPoint2f prevPts,
  VecPoint2f nextPts, {
  VecUChar? status,
  VecF32? err,
  (int, int) winSize = (21, 21),
  int maxLevel = 3,
  (int, int, double) criteria = (TERM_COUNT + TERM_EPS, 30, 1e-4),
  int flags = 0,
  double minEigThreshold = 1e-4,
}) {
  status ??= VecUChar();
  err ??= VecF32();
  cvRun(
    () => cvideo.cv_calcOpticalFlowPyrLK_1(
      prevImg.ref,
      nextImg.ref,
      prevPts.ref,
      nextPts.ptr,
      status!.ptr,
      err!.ptr,
      winSize.cvd.ref,
      maxLevel,
      criteria.toTermCriteria().ref,
      flags,
      minEigThreshold,
      ffi.nullptr,
    ),
  );
  return (nextPts, status, err);
}

/// FindTransformECC finds the geometric transform (warp) between two images in terms of the ECC criterion.
///
/// For futther details, please see:
/// https://docs.opencv.org/4.x/dc/d6b/group__video__track.html#ga1aa357007eaec11e9ed03500ecbcbe47
(double ret, Mat warpMatrix) findTransformECC(
  InputArray templateImage,
  InputArray inputImage,
  InputOutputArray warpMatrix,
  int motionType,
  (int, int, double) criteria,
  InputArray inputMask,
  int gaussFiltSize,
) {
  final p = calloc<ffi.Double>();
  cvRun(
    () => cvideo.cv_findTransformECC(
      templateImage.ref,
      inputImage.ref,
      warpMatrix.ref,
      motionType,
      criteria.toTermCriteria().ref,
      inputMask.ref,
      gaussFiltSize,
      p,
      ffi.nullptr,
    ),
  );
  final rval = (p.value, warpMatrix);
  calloc.free(p);
  return rval;
}

/// Tracker is the base interface for object tracking.
///
/// see: https://docs.opencv.org/master/d0/d0a/classcv_1_1Tracker.html
class TrackerMIL extends CvStruct<cvg.TrackerMIL> {
  TrackerMIL(cvg.TrackerMILPtr ptr, [bool attach = true]) : super.fromPointer(ptr) {
    if (attach) {
      finalizer.attach(this, ptr.cast(), detach: this);
    }
  }
  factory TrackerMIL.create() {
    final p = calloc<cvg.TrackerMIL>();
    cvRun(() => cvideo.cv_TrackerMIL_create(p));
    return TrackerMIL(p);
  }

  void init(InputArray image, Rect boundingBox) {
    cvAssert(boundingBox.x >= 0, "boundingBox.x must be >= 0");
    cvAssert(boundingBox.y >= 0, "boundingBox.y must be >= 0");
    cvAssert(
      boundingBox.right <= image.cols,
      "boundingBox.right=${boundingBox.right} must be <= image.cols=${image.cols}",
    );
    cvAssert(
      boundingBox.bottom <= image.rows,
      "boundingBox.bottom=${boundingBox.bottom} must be <= image.rows=${image.rows}",
    );

    cvRun(() => cvideo.cv_TrackerMIL_init(ref, image.ref, boundingBox.ref, ffi.nullptr));
  }

  /// Update the tracker, find the new most likely bounding box for the target.
  /// https://docs.opencv.org/4.x/d0/d0a/classcv_1_1Tracker.html#a92d2012f576e6c06eb2e257d110a6529
  (bool, Rect) update(Mat img) {
    final bBox = calloc<cvg.CvRect>();
    final p = calloc<ffi.Bool>();
    cvRun(() => cvideo.cv_TrackerMIL_update(ref, img.ref, bBox, p, ffi.nullptr));
    final rval = (p.value, Rect.fromPointer(bBox));
    calloc.free(p);
    return rval;
  }

  static final finalizer = OcvFinalizer<cvg.TrackerMILPtr>(cvideo.addresses.cv_TrackerMIL_close);

  void dispose() {
    finalizer.detach(this);
    cvideo.cv_TrackerMIL_close(ptr);
  }

  @override
  cvg.TrackerMIL get ref => ptr.ref;
}

/// KalmanFilter implements a standard Kalman filter http://en.wikipedia.org/wiki/Kalman_filter.
/// However, you can modify transitionMatrix, controlMatrix, and measurementMatrix
/// to get an extended Kalman filter functionality.
///
/// For further details, please see:
/// https://docs.opencv.org/4.6.0/dd/d6a/classcv_1_1KalmanFilter.html
class KalmanFilter extends CvStruct<cvg.KalmanFilter> {
  KalmanFilter(cvg.KalmanFilterPtr ptr, [bool attach = true]) : super.fromPointer(ptr) {
    if (attach) {
      finalizer.attach(this, ptr.cast(), detach: this);
    }
  }

  factory KalmanFilter.create(
    int dynamParams,
    int measureParams, {
    int controlParams = 0,
    int type = MatType.CV_32F,
  }) {
    final p = calloc<cvg.KalmanFilter>();
    cvRun(() => cvideo.cv_KalmanFilter_create(dynamParams, measureParams, controlParams, type, p));
    return KalmanFilter(p);
  }

  Mat correct(Mat measurement) {
    final dst = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_correct(ref, measurement.ref, dst.ptr, ffi.nullptr));
    return dst;
  }

  Mat predict({Mat? control}) {
    final dst = Mat.empty();
    control == null
        ? cvRun(() => cvideo.cv_KalmanFilter_predict(ref, dst.ptr, ffi.nullptr))
        : cvRun(() => cvideo.cv_KalmanFilter_predict_1(ref, control.ref, dst.ptr, ffi.nullptr));
    return dst;
  }

  void init(
    int dynamParams,
    int measureParams, {
    int controlParams = 0,
    int type = MatType.CV_32F,
  }) {
    cvRun(
      () => cvideo.cv_KalmanFilter_init_1(ref, dynamParams, measureParams, controlParams, type, ffi.nullptr),
    );
  }

  @override
  cvg.KalmanFilter get ref => ptr.ref;
  static final finalizer = OcvFinalizer<cvg.KalmanFilterPtr>(cvideo.addresses.cv_KalmanFilter_close);

  void dispose() {
    finalizer.detach(this);
    cvideo.cv_KalmanFilter_close(ptr);
  }

  // corrected state (x(k)): x(k)=x'(k)+K(k)*(z(k)-H*x'(k))
  Mat get statePost {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_statePost(ref, p.ptr));
    return p;
  }

  set statePost(Mat state) {
    cvRun(() => cvideo.cv_KalmanFilter_set_statePost(ref, state.ref));
  }

  Mat get statePre {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_statePre(ref, p.ptr));
    return p;
  }

  set statePre(Mat state) {
    cvRun(() => cvideo.cv_KalmanFilter_set_statePre(ref, state.ref));
  }

  Mat get transitionMatrix {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_transitionMatrix(ref, p.ptr));
    return p;
  }

  set transitionMatrix(Mat m) {
    cvRun(() => cvideo.cv_KalmanFilter_set_transitionMatrix(ref, m.ref));
  }

  Mat get temp1 {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_temp1(ref, p.ptr));
    return p;
  }

  Mat get temp2 {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_temp2(ref, p.ptr));
    return p;
  }

  Mat get temp3 {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_temp3(ref, p.ptr));
    return p;
  }

  Mat get temp4 {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_temp4(ref, p.ptr));
    return p;
  }

  Mat get temp5 {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_temp5(ref, p.ptr));
    return p;
  }

  Mat get processNoiseCov {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_processNoiseCov(ref, p.ptr));
    return p;
  }

  set processNoiseCov(Mat m) {
    cvRun(() => cvideo.cv_KalmanFilter_set_processNoiseCov(ref, m.ref));
  }

  Mat get measurementNoiseCov {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_measurementNoiseCov(ref, p.ptr));
    return p;
  }

  set measurementNoiseCov(Mat m) {
    cvRun(() => cvideo.cv_KalmanFilter_set_measurementNoiseCov(ref, m.ref));
  }

  Mat get measurementMatrix {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_measurementMatrix(ref, p.ptr));
    return p;
  }

  set measurementMatrix(Mat m) {
    cvRun(() => cvideo.cv_KalmanFilter_set_measurementMatrix(ref, m.ref));
  }

  Mat get gain {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_gain(ref, p.ptr));
    return p;
  }

  set gain(Mat m) {
    cvRun(() => cvideo.cv_KalmanFilter_set_gain(ref, m.ref));
  }

  Mat get errorCovPre {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_errorCovPre(ref, p.ptr));
    return p;
  }

  set errorCovPre(Mat m) {
    cvRun(() => cvideo.cv_KalmanFilter_set_errorCovPre(ref, m.ref));
  }

  Mat get errorCovPost {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_errorCovPost(ref, p.ptr));
    return p;
  }

  set errorCovPost(Mat m) {
    cvRun(() => cvideo.cv_KalmanFilter_set_errorCovPost(ref, m.ref));
  }

  Mat get controlMatrix {
    final p = Mat.empty();
    cvRun(() => cvideo.cv_KalmanFilter_get_controlMatrix(ref, p.ptr));
    return p;
  }

  set controlMatrix(Mat m) {
    cvRun(() => cvideo.cv_KalmanFilter_set_controlMatrix(ref, m.ref));
  }
}

File: src/dartcv/video/video.cpp
/*
    This file is originally from gocv project
    Licensed: Apache 2.0 license. Copyright (c) 2017-2021 The Hybrid Group.

    Modified by Rainyl.
    Licensed: Apache 2.0 license. Copyright (c) 2024 Rainyl.
*/

#include "dartcv/video/video.h"
#include "dartcv/core/vec.hpp"

#include <iostream>

CvStatus* cv_BackgroundSubtractorMOG2_create(BackgroundSubtractorMOG2* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Ptr<cv::BackgroundSubtractorMOG2>(cv::createBackgroundSubtractorMOG2());
    END_WRAP
}
CvStatus* cv_BackgroundSubtractorMOG2_create_1(
    int history, double varThreshold, bool detectShadows, BackgroundSubtractorMOG2* rval
) {
    BEGIN_WRAP
    rval->ptr = new cv::Ptr<cv::BackgroundSubtractorMOG2>(
        cv::createBackgroundSubtractorMOG2(history, varThreshold, detectShadows)
    );
    END_WRAP
}
void cv_BackgroundSubtractorMOG2_close(BackgroundSubtractorMOG2Ptr self) {
    self->ptr->reset();
    CVD_FREE(self);
}

CvStatus* cv_BackgroundSubtractorMOG2_apply(
    BackgroundSubtractorMOG2 self, Mat src, Mat dst, CvCallback_0 callback
) {
    BEGIN_WRAP(CVDEREF(self))->apply(CVDEREF(src), CVDEREF(dst));
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

CvStatus* cv_BackgroundSubtractorKNN_create(BackgroundSubtractorKNN* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Ptr<cv::BackgroundSubtractorKNN>(cv::createBackgroundSubtractorKNN());
    END_WRAP
}
CvStatus* cv_BackgroundSubtractorKNN_create_1(
    int history, double dist2Threshold, bool detectShadows, BackgroundSubtractorKNN* rval
) {
    BEGIN_WRAP
    rval->ptr = new cv::Ptr<cv::BackgroundSubtractorKNN>(
        cv::createBackgroundSubtractorKNN(history, dist2Threshold, detectShadows)
    );
    END_WRAP
}
void cv_BackgroundSubtractorKNN_close(BackgroundSubtractorKNNPtr self) {
    self->ptr->reset();
    CVD_FREE(self);
}

CvStatus* cv_BackgroundSubtractorKNN_apply(
    BackgroundSubtractorKNN self, Mat src, Mat dst, CvCallback_0 callback
) {
    BEGIN_WRAP(CVDEREF(self))->apply(CVDEREF(src), CVDEREF(dst));
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

CvStatus* cv_calcOpticalFlowPyrLK(
    Mat prevImg,
    Mat nextImg,
    VecPoint2f prevPts,
    VecPoint2f* nextPts,
    VecUChar* status,
    VecF32* err,
    CvCallback_0 callback
) {
    BEGIN_WRAP
    cv::calcOpticalFlowPyrLK(
        CVDEREF(prevImg), CVDEREF(nextImg), CVDEREF(prevPts), CVDEREF_P(nextPts), CVDEREF_P(status), CVDEREF_P(err)
    );
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

CvStatus* cv_calcOpticalFlowPyrLK_1(
    Mat prevImg,
    Mat nextImg,
    VecPoint2f prevPts,
    VecPoint2f* nextPts,
    VecUChar* status,
    VecF32* err,
    CvSize winSize,
    int maxLevel,
    TermCriteria criteria,
    int flags,
    double minEigThreshold,
    CvCallback_0 callback
) {
    BEGIN_WRAP
    auto tc = cv::TermCriteria(criteria.type, criteria.maxCount, criteria.epsilon);
    cv::calcOpticalFlowPyrLK(
        CVDEREF(prevImg),
        CVDEREF(nextImg),
        CVDEREF(prevPts),
        CVDEREF_P(nextPts),
        CVDEREF_P(status),
        CVDEREF_P(err),
        cv::Size(winSize.width, winSize.height),
        maxLevel,
        tc,
        flags,
        minEigThreshold
    );
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}
CvStatus* cv_calcOpticalFlowFarneback(
    Mat prevImg,
    Mat nextImg,
    Mat flow,
    double pyrScale,
    int levels,
    int winsize,
    int iterations,
    int polyN,
    double polySigma,
    int flags,
    CvCallback_0 callback
) {
    BEGIN_WRAP
    cv::calcOpticalFlowFarneback(
        CVDEREF(prevImg),
        CVDEREF(nextImg),
        CVDEREF(flow),
        pyrScale,
        levels,
        winsize,
        iterations,
        polyN,
        polySigma,
        flags
    );
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

CvStatus* cv_findTransformECC(
    Mat templateImage,
    Mat inputImage,
    Mat warpMatrix,
    int motionType,
    TermCriteria criteria,
    Mat inputMask,
    int gaussFiltSize,
    double* rval,
    CvCallback_0 callback
) {
    BEGIN_WRAP
    auto tc = cv::TermCriteria(criteria.type, criteria.maxCount, criteria.epsilon);
    *rval = cv::findTransformECC(
        CVDEREF(templateImage),
        CVDEREF(inputImage),
        CVDEREF(warpMatrix),
        motionType,
        tc,
        CVDEREF(inputMask),
        gaussFiltSize
    );
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

CvStatus* cv_TrackerMIL_init(TrackerMIL self, Mat image, CvRect bbox, CvCallback_0 callback) {
    BEGIN_WRAP
    auto rect = cv::Rect(bbox.x, bbox.y, bbox.width, bbox.height);
    (CVDEREF(self))->init(CVDEREF(image), rect);
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}
CvStatus* cv_TrackerMIL_update(
    TrackerMIL self, Mat image, CvRect* boundingBox, bool* rval, CvCallback_0 callback
) {
    BEGIN_WRAP
    cv::Rect bb;
    *rval = (CVDEREF(self))->update(CVDEREF(image), bb);
    *boundingBox = {bb.x, bb.y, bb.width, bb.height};
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

CvStatus* cv_TrackerMIL_create(TrackerMIL* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Ptr<cv::TrackerMIL>(cv::TrackerMIL::create());
    END_WRAP
}
void cv_TrackerMIL_close(TrackerMILPtr self) {
    self->ptr->reset();
    CVD_FREE(self);
}

CvStatus* cv_KalmanFilter_create(
    int dynamParams, int measureParams, int controlParams, int type, KalmanFilter* rval
) {
    BEGIN_WRAP
    rval->ptr = new cv::KalmanFilter(dynamParams, measureParams, controlParams, type);
    END_WRAP
}
void cv_KalmanFilter_close(KalmanFilterPtr self) {
    CVD_FREE(self);
}

CvStatus* cv_KalmanFilter_init(
    KalmanFilter self, int dynamParams, int measureParams, CvCallback_0 callback
) {
    BEGIN_WRAP
    self.ptr->init(dynamParams, measureParams);
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}
CvStatus* cv_KalmanFilter_init_1(
    KalmanFilter self,
    int dynamParams,
    int measureParams,
    int controlParams,
    int type,
    CvCallback_0 callback
) {
    BEGIN_WRAP
    self.ptr->init(dynamParams, measureParams, controlParams, type);
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}
CvStatus* cv_KalmanFilter_predict(KalmanFilter self, Mat* rval, CvCallback_0 callback) {
    BEGIN_WRAP
    auto result = self.ptr->predict();
    rval->ptr = new cv::Mat(result);
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}
CvStatus* cv_KalmanFilter_predict_1(
    KalmanFilter self, Mat control, Mat* rval, CvCallback_0 callback
) {
    BEGIN_WRAP
    auto result = self.ptr->predict(CVDEREF(control));
    rval->ptr = new cv::Mat(result);
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}
CvStatus* cv_KalmanFilter_correct(
    KalmanFilter self, Mat measurement, Mat* rval, CvCallback_0 callback
) {
    BEGIN_WRAP
    auto result = self.ptr->correct(CVDEREF(measurement));
    rval->ptr = new cv::Mat(result);
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

CvStatus* cv_KalmanFilter_get_statePre(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->statePre);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_statePost(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->statePost);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_transitionMatrix(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->transitionMatrix);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_controlMatrix(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->controlMatrix);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_measurementMatrix(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->measurementMatrix);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_processNoiseCov(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->processNoiseCov);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_measurementNoiseCov(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->measurementNoiseCov);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_errorCovPre(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->errorCovPre);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_gain(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->gain);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_errorCovPost(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->errorCovPost);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_temp1(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->temp1);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_temp2(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->temp2);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_temp3(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->temp3);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_temp4(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->temp4);
    END_WRAP
}
CvStatus* cv_KalmanFilter_get_temp5(KalmanFilter self, Mat* rval) {
    BEGIN_WRAP
    rval->ptr = new cv::Mat(self.ptr->temp5);
    END_WRAP
}

CvStatus* cv_KalmanFilter_set_statePre(KalmanFilter self, Mat statePre) {
    BEGIN_WRAP
    self.ptr->statePre = CVDEREF(statePre);
    END_WRAP
}
CvStatus* cv_KalmanFilter_set_statePost(KalmanFilter self, Mat statePost) {
    BEGIN_WRAP
    self.ptr->statePost = CVDEREF(statePost);
    END_WRAP
}
CvStatus* cv_KalmanFilter_set_transitionMatrix(KalmanFilter self, Mat transitionMatrix) {
    BEGIN_WRAP
    self.ptr->transitionMatrix = CVDEREF(transitionMatrix);
    END_WRAP
}
CvStatus* cv_KalmanFilter_set_controlMatrix(KalmanFilter self, Mat controlMatrix) {
    BEGIN_WRAP
    self.ptr->controlMatrix = CVDEREF(controlMatrix);
    END_WRAP
}
CvStatus* cv_KalmanFilter_set_measurementMatrix(KalmanFilter self, Mat measurementMatrix) {
    BEGIN_WRAP
    self.ptr->measurementMatrix = CVDEREF(measurementMatrix);
    END_WRAP
}
CvStatus* cv_KalmanFilter_set_processNoiseCov(KalmanFilter self, Mat processNoiseCov) {
    BEGIN_WRAP
    self.ptr->processNoiseCov = CVDEREF(processNoiseCov);
    END_WRAP
}
CvStatus* cv_KalmanFilter_set_measurementNoiseCov(KalmanFilter self, Mat measurementNoiseCov) {
    BEGIN_WRAP
    self.ptr->measurementNoiseCov = CVDEREF(measurementNoiseCov);
    END_WRAP
}
CvStatus* cv_KalmanFilter_set_errorCovPre(KalmanFilter self, Mat errorCovPre) {
    BEGIN_WRAP
    self.ptr->errorCovPre = CVDEREF(errorCovPre);
    END_WRAP
}
CvStatus* cv_KalmanFilter_set_gain(KalmanFilter self, Mat gain) {
    BEGIN_WRAP
    self.ptr->gain = CVDEREF(gain);
    END_WRAP
}
CvStatus* cv_KalmanFilter_set_errorCovPost(KalmanFilter self, Mat errorCovPost) {
    BEGIN_WRAP
    self.ptr->errorCovPost = CVDEREF(errorCovPost);
    END_WRAP
}
